File: orchestra/policy.py
"""Pure scheduling and activity policy."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Mapping

from .models import Activity


@dataclass(frozen=True, slots=True)
class Candidate:
    priority: int
    key: str
    kind: str


def fresh_activities(
    activities: Iterable[Activity],
    seen: Mapping[str, float],
    bot_identities: frozenset[str],
    require_directed: bool,
) -> list[Activity]:
    """Return remote activity that is new, not self-authored, and addressable."""

    return [
        activity
        for activity in activities
        if activity.author not in bot_identities
        and activity.cursor > float(seen.get(activity.stream, 0.0))
        and (
            activity.directed is not False
            or not require_directed
        )
    ]


def advance_seen(
    seen: Mapping[str, float], activities: Iterable[Activity]
) -> dict[str, float]:
    advanced = dict(seen)
    for activity in activities:
        advanced[activity.stream] = max(
            float(advanced.get(activity.stream, 0.0)), activity.cursor
        )
    return advanced


def activity_prompt(activities: Iterable[Activity]) -> str:
    """Build a pointer-only prompt grouped by remote location."""

    groups: dict[tuple[str, str, str, str, str], int] = {}
    order: dict[tuple[str, str, str, str, str], float] = {}
    for activity in activities:
        key = (
            activity.ref,
            activity.thread,
            activity.author,
            activity.kind,
            activity.url,
        )
        groups[key] = groups.get(key, 0) + 1
        order[key] = min(order.get(key, activity.cursor), activity.cursor)

    lines = ["New activity is waiting at the source:"]
    for key in sorted(groups, key=lambda value: order[value]):
        ref, thread, author, kind, url = key
        where = ref or "task"
        if thread:
            where += f' / thread "{thread}"'
        who = f" from {author}" if author else ""
        lines.append(f"- {where}: {groups[key]} {kind} event(s){who}\n  {url}")
    lines.append("\nRead the authoritative content there, then act on it.")
    return "\n".join(lines)


def sort_candidates(candidates: Iterable[Candidate]) -> list[Candidate]:
    kind_order = {"retry": 0, "activity": 1, "bootstrap": 2, "claim": 2}
    return sorted(
        candidates,
        key=lambda candidate: (
            kind_order[candidate.kind],
            candidate.priority,
            candidate.key,
        ),
    )

File: orchestra/policy.ts
import type { Activity } from "./models.js";
import { lookup } from "./json.js";

export interface Candidate {
  readonly priority: number;
  readonly key: string;
  readonly kind: "retry" | "activity" | "bootstrap" | "claim";
}
export function freshActivities(
  activities: readonly Activity[],
  seen: Readonly<Record<string, number>>,
  botIdentities: ReadonlySet<string>,
  requireDirected: boolean,
): Activity[] {
  return activities.filter(
    (event) =>
      !botIdentities.has(event.author) &&
      event.cursor > (lookup(seen, event.stream) ?? 0) &&
      (event.directed !== false || !requireDirected),
  );
}
export function advanceSeen(
  seen: Readonly<Record<string, number>>,
  activities: readonly Activity[],
): Record<string, number> {
  const advanced = Object.assign(
    Object.create(null) as Record<string, number>,
    seen,
  );
  for (const event of activities)
    advanced[event.stream] = Math.max(
      advanced[event.stream] ?? 0,
      event.cursor,
    );
  return { ...advanced };
}
export function activityPrompt(activities: readonly Activity[]): string {
  const groups = new Map<
    string,
    { event: Activity; count: number; cursor: number }
  >();
  for (const event of activities) {
    const key = JSON.stringify([
      event.ref ?? "",
      event.thread ?? "",
      event.author,
      event.kind,
      event.url,
    ]);
    const previous = groups.get(key);
    groups.set(key, {
      event,
      count: (previous?.count ?? 0) + 1,
      cursor: Math.min(previous?.cursor ?? event.cursor, event.cursor),
    });
  }
  const lines = ["New activity is waiting at the source:"];
  for (const { event, count } of [...groups.values()].sort(
    (a, b) => a.cursor - b.cursor,
  )) {
    const where =
      (event.ref || "task") +
      (event.thread ? ` / thread "${event.thread}"` : "");
    lines.push(
      `- ${where}: ${count} ${event.kind} event(s)${event.author ? ` from ${event.author}` : ""}\n  ${event.url}`,
    );
  }
  return [
    ...lines,
    "\nRead the authoritative content there, then act on it.",
  ].join("\n");
}
export function sortCandidates(candidates: readonly Candidate[]): Candidate[] {
  const order = { retry: 0, activity: 1, bootstrap: 2, claim: 2 };
  return [...candidates].sort(
    (a, b) =>
      order[a.kind] - order[b.kind] ||
      a.priority - b.priority ||
      (a.key < b.key ? -1 : a.key > b.key ? 1 : 0),
  );
}
